Document TypeBindingOptions (#281)
Starcounter/Docs: docs/database-classes.md

});
```

## Discovery

Starcounter automatically discovers all database classes in the application assembly, which does not include all referenced assemblies. Manual configuration is required to use database classes from referenced assemblies.

> ### Referenced assembly

Reviewer: Personally, I would not use headings for this illustration. Instead, describe the separation in body text.

 ```suggestion
### Example

The `Something` database class definition below is contained in an assembly that our app references:

```

```cs
[Database]
public abstract class Something
{
}
```

> ### Application assembly

Reviewer: Personally, I would not use headings for this illustration. Instead, describe the separation in body text.

 ```suggestion

### Example

The `Something` database class definition below is contained in an assembly that our app references:

```cs
[Database]
public abstract class Something
{
}
```

And here's the code we need to have in our application for it to detect the database class definition above:
```

> ### Application assembly

Reviewer: ```suggestion
### Application assembly
```
And here's the code we need to have in our application for it to detect the database class definition above:

> ### Application assembly

Reviewer: ```suggestion
And here's the code we need to have in our application for it to detect the database class definition above:
```

> 

Reviewer: ```suggestion
And here's the code we need to have in our application for it to detect the database class definition above:

```

```cs
using var services = new ServiceCollection()
    .AddStarcounter($"Database=./.database/test;OpenMode=CreateIfNotExists;StartMode=StartIfNotRunning;StopMode=IfWeStarted")
    .Configure<Starcounter.Database.Binding.TypeBindingOptions>(o =>
    {
        Type[] extraDatabaseTypes = typeof(Something) // Getting type of a database class.
            .Assembly // Getting assembly which defines the type.
            .GetTypes() // Getting all types in the assembly.

>             .Where(x => x.GetCustomAttributes(typeof(DatabaseAttribute), true).Any()) // Filtering the types by the `[Database]` attribute.

Reviewer: ```suggestion
            .Where(x => x.GetCustomAttribute<DatabaseAttribute>(inherit: true) != null) // Filtering the types by the `DatabaseAttribute` attribute.
```

I would prefer the generic method finding the first decoration, and comparing with null.

});
```

## Discovery

Starcounter automatically discovers all database classes in the application assembly, which does not include all referenced assemblies. Manual configuration is required to use database classes from referenced assemblies.

### Example

The `Something` database class definition below is contained in an assembly that our app references:

```cs
[Database]
public abstract class Something
{
}
```

And here's the code we need to have in our application for it to detect the database class definition above, as well as any other database types contained in the same assembly:

```cs
using var services = new ServiceCollection()
    .AddStarcounter($"Database=./.database/test;OpenMode=CreateIfNotExists;StartMode=StartIfNotRunning;StopMode=IfWeStarted")
    .Configure<Starcounter.Database.Binding.TypeBindingOptions>(o =>
    {
        Type[] extraDatabaseTypes = typeof(Something) // Getting type of a database class.
            .Assembly // Getting assembly which defines the type.

>             .GetTypes() // Getting all types in the assembly.

Reviewer 2: I think we should use the `ExportedTypes` property.

>             .Where(x => x.GetCustomAttribute<DatabaseAttribute>(inherit: true) != null) // Filtering the types by the `DatabaseAttribute` attribute.

Reviewer 2: ```suggestion
            .Where(t => t.IsDatabaseType()) // Filtering the types by the `DatabaseAttribute` attribute.
```

            .ToArray();

        // The `o.DefaultTypes` property contains all automatically discovered database classes.
        // The `o.Types` property defines which classes will be treated as database classes.
        // It shall contain all automatically discovered database classes plus extra classes from the referenced assembly.

>         o.Types = o.DefaultTypes.Concat(extraDatabaseTypes).ToArray();

Reviewer 2: Why these 2 `ToArray()` calls? Better to just use `IEnumerable`.

Author: The first one is because I want to use explicit type, and `Type[]` looks simpler / better than `IEnumerable<Type>`.

The second one is because `IEnumerable` is lazy, while `.ToArray` will explicitly determine the list of types at the moment of assignement.

Reviewer 2: For me, I always want us to use `IEnumerable` when there is no absolute use something else. First, it better illustrate the way we intend to use the result - we are just enumerating it. Second, I think it's more standard.

As for the lazyness, it's for the good. If the `Types` parameter is never consumed, iteration don't need to take place.

But nevermind here, if you feel it's clearer in documentation. 👍

});
```

## Discovery

Starcounter automatically discovers all database classes in the application assembly, which does not include all referenced assemblies. Manual configuration is required to use database classes from referenced assemblies.

### Example

The `Something` database class definition below is contained in an assembly that our app references:

> 

Reviewer: ```suggestion
```

```cs
[Database]
public abstract class Something
{
}
```

> And here's the code we need to have in our application for it to detect the database class definition above:

Reviewer: ```suggestion
And here's the code we need to have in our application for it to detect the database class definition above, as well as any other database types contained in the same assembly:
```

should probably add that to not confuse people ;)
